test/lang/python/smoketest.py

```diff
@@ -7,14 +7,99 @@
+                self.assertEqual(diff, 1)
+                self.assertEqual(a, 1)
+                self.assertEqual(b, "a")
+                self.assertEqual(cur.fetchone(), None)
+
+                # Insert another row from another connection to simulate an
+                # update arriving.
+                with psycopg2.connect(MATERIALIZED_URL) as conn2:
+                    with conn2.cursor() as cur2:
+                        cur2.execute("INSERT INTO psycopg2_tail VALUES (2, 'b')")
+
+                # Validate the new row, again ignoring the timestamp column.
+                cur.execute("FETCH ALL cur")
+                (ts, diff, a, b) = cur.fetchone()
+                self.assertEqual(diff, 1)
+                self.assertEqual(a, 2)
+                self.assertEqual(b, "b")
+                self.assertEqual(cur.fetchone(), None)
+
+    def test_psycopg3_tail(self):
+        """Test tail with psycopg3 via its new binary COPY decoding support."""
+        conn = psycopg3.connect(MATERIALIZED_URL)
+        with conn.cursor() as cur:
+            # Create a table with one row of data.
+            cur.execute("CREATE TABLE psycopg3_tail (a int, b text)")
+            cur.execute("INSERT INTO psycopg3_tail VALUES (1, 'a')")
+
+            # Start a tail using the binary copy protocol.
+            with cur.copy(
+                "COPY (TAIL psycopg3_tail) TO STDOUT (FORMAT BINARY)"
+            ) as copy:
+                copy.set_types(
+                    [
+                        builtins["numeric"].oid,  # timestamp
+                        builtins["int8"].oid,  # diff
+                        builtins["int4"].oid,  # a column
+                        builtins["text"].oid,  # b column
+                    ]
+                )
+
+                # Validate the first row, but ignore the timestamp column.
+                (ts, diff, a, b) = copy.read_row()
+                self.assertEqual(diff, 1)
+                self.assertEqual(a, 1)
+                self.assertEqual(b, "a")
+
+                # Insert another row from another connection to simulate an
+                # update arriving.
+                with psycopg3.connect(MATERIALIZED_URL) as conn2:
+                    with conn2.cursor() as cur2:
+                        cur2.execute("INSERT INTO psycopg3_tail VALUES (2, 'b')")
+
+                # Validate the new row, again ignoring the timestamp column.
+                (ts, diff, a, b) = copy.read_row()
+                self.assertEqual(diff, 1)
+                self.assertEqual(a, 2)
+                self.assertEqual(b, "b")
+
+                # The tail won't end until we send a cancel request.
+                conn.cancel()
+                self.assertEqual(copy.read_row(), None)
+
+        # TODO(benesch): it should be possible to use a `with` context manager
+        # here, but doing so triggers a commit when the context manager exits,
+        # and that commit fails with a libpq "operation already in progress"
```

Comment:
Thank you for reporting the issue, I'll look into it. Please note that you can use an autocommit connection to avoid the COMMIT command to be sent.